### DISCUSSION TITLE
Unable to launch Vaultwarden after declaring keyfiles in Rocket_TLS

### DISCUSSION THREAD

**Discussion by bmpreston** [dani-garcia/vaultwarden#4381] [Q&A] [answered]
Good evening,

I've been toying forever with trying to get a single HTTPS instance of Vaultwarden and I'm running this on an Unraid 6.12.8 box.  I have installed SWAG and received the SSL PEM files.  These files are now declared in /ssl:

/ssl<-->/mnt/user/appdata/swag/keys/letsencrypt/

I can see these files in the container.  They did map correctly and they exist.  

The problem is when I declare the ROCKET_TLS variable.  This causes the container to constantly restart.  That is currently:

'{certs="/ssl/fullchain.pem",key="/ssl/privkey.pem"}'

That is the EXACT variable I put against ROCKET_TLS.  When I remove the variable, it will start and launch.  

I have spent hours trying to determine the cause.  Any help on this?

**Comment by stefan0xC** [comment]
What do the logs say? If it's something like this try to remove the `'`.
```
[2024-02-26 06:35:08.113][rocket::config::config][ERROR] Failed to extract valid configuration.
[2024-02-26 06:35:08.122][rocket::config::config::_][ERROR] invalid type: found string "'{certs=\"/ssl/fullchain.pem\",key=\"/ssl/privkey.pem\"}'", expected struct TlsConfig
...
```

**Comment by bmpreston** [comment]
Problem is I don’t know how to pull the logs since it’s short cycling.

If you have suggestions, please let me know.

Thank you

**Reply by BlackDex** [reply]
`docker logs -f vaultwarden`, or whatever you called your container. Or, with docker compose, do not use `-d` it will then run in the foreground.

**Comment by bmpreston** [comment]
I’ll run this shortly.

My problem is I’m on Unraid. Those logs in the Docker tab are only
accessible when the app is running.

I’ll ssh and check them myself soon.

**Comment by bmpreston** [comment]
> [2024-02-26 13:44:48.507][rocket::config::config][ERROR] Failed to extract valid configuration.
[2024-02-26 13:44:48.507][rocket::config::config::_][ERROR] invalid type: found string "'{certs=\"/ssl/fullchain.pem\",key=\"/ssl/privkey.pem\"}'", expected struct TlsConfig
[2024-02-26 13:44:48.507][rocket::config::config::_][INFO] for key TLS
[2024-02-26 13:44:48.507][rocket::config::config::_][INFO] in `ROCKET_` environment variable(s)
[2024-02-26 13:44:48.508][panic][ERROR] thread 'main' panicked at 'aborting due to configuration error(s)': /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rocket-0.5.0/src/config/config.rs:589
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: vaultwarden::launch_rocket::{{closure}}
   7: vaultwarden::main::{{closure}}
   8: vaultwarden::main
   9: std::sys_common::backtrace::__rust_begin_short_backtrace
  10: std::rt::lang_start::{{closure}}
  11: std::rt::lang_start
  12: <unknown>
  13: __libc_start_main
  14: _start

![Screenshot](https://github.com/dani-garcia/vaultwarden/assets/29456939/b1df6bd9-4805-44d8-a18e-a98ebaf27dab)

**Reply by bmpreston** [reply]
I wanted to clarify, the certs ARE in the /ssl folder:

<img width="1043" alt="Screenshot" src="https://github.com/dani-garcia/vaultwarden/assets/29456939/78c00354-13e9-4c70-9204-d8bcecfb23f8">

**Reply by BlackDex** [reply]
Looks like the quotes are not correctly
Either use one type.of quotes and escape them where needed, or use two types.

`ROCKET_TLS="{certs=’/path/to.crt',key='/path/to.key'}"`
Also make sure those files are not symlinks to sometimes outside the mounted path.

**Reply by BlackDex** [reply]
What does a `ls -al` look like?

**Reply by bmpreston** [reply]
I had to readlink them.  I found them. They show up as 

/mnt/user/appdata/swag/etc/letsencrypt/archive/foobar.net

**Reply by bmpreston** [reply]
`
[WARNING] The following environment variables are being overridden by the config.json file.
[WARNING] Please use the admin panel to make changes to them:
[WARNING] SIGNUPS_ALLOWED, INVITATIONS_ALLOWED, ADMIN_TOKEN

[NOTICE] You are using a plain text `ADMIN_TOKEN` which is insecure.
Please generate a secure Argon2 PHC string by using `vaultwarden hash` or `argon2`.
See: https://github.com/dani-garcia/vaultwarden/wiki/Enabling-admin-page#secure-the-admin_token

[2024-02-26 14:20:02.056][rocket::config::config][ERROR] Failed to extract valid configuration.
[2024-02-26 14:20:02.056][rocket::config::config::_][ERROR] invalid type: found string "{certs=’/ssl/cert1.pem',key='/ssl/fullchain1.pem'}", expected struct TlsConfig
[2024-02-26 14:20:02.056][rocket::config::config::_][INFO] for key TLS
[2024-02-26 14:20:02.056][rocket::config::config::_][INFO] in `ROCKET_` environment variable(s)
[2024-02-26 14:20:02.056][panic][ERROR] thread 'main' panicked at 'aborting due to configuration error(s)': /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rocket-0.5.0/src/config/config.rs:589
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: vaultwarden::launch_rocket::{{closure}}
   7: vaultwarden::main::{{closure}}
   8: vaultwarden::main
   9: std::sys_common::backtrace::__rust_begin_short_backtrace
  10: std::rt::lang_start::{{closure}}
  11: std::rt::lang_start
  12: <unknown>
  13: __libc_start_main
  14: _start

**Reply by BlackDex** [reply]
Well, that's not gonna work then.
Links either hard or sym do not work that well.
They need to be the actual files.

**Reply by BlackDex** [reply]
Then use no quotes around the whole string, see what that does.

**Reply by bmpreston** [reply]
`[2024-02-26 14:23:08.183][rocket::config::config][ERROR] Failed to extract valid configuration.
[2024-02-26 14:23:08.183][rocket::config::config::_][ERROR] invalid type: found string "{certs=’/ssl/cert1.pem',key='/ssl/fullchain1.pem'}", expected struct TlsConfig
[2024-02-26 14:23:08.183][rocket::config::config::_][INFO] for key TLS
[2024-02-26 14:23:08.183][rocket::config::config::_][INFO] in `ROCKET_` environment variable(s)
[2024-02-26 14:23:08.183][panic][ERROR] thread 'main' panicked at 'aborting due to configuration error(s)': /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/rocket-0.5.0/src/config/config.rs:589
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: vaultwarden::launch_rocket::{{closure}}
   7: vaultwarden::main::{{closure}}
   8: vaultwarden::main
   9: std::sys_common::backtrace::__rust_begin_short_backtrace
  10: std::rt::lang_start::{{closure}}
  11: std::rt::lang_start
  12: <unknown>
  13: __libc_start_main
  14: _start`

**Reply by bmpreston** [reply]
<img width="543" alt="Screenshot" src="https://github.com/dani-garcia/vaultwarden/assets/29456939/72ccfd53-8ebb-4aac-8874-75861cc985e9">

The files DO exist

**Reply by BlackDex** [reply]
That doesn't show if they are symlinks or not.
Run cat on them to be sure.

**Reply by stefan0xC** [reply, marked as answer]
That's not the issue. Cf. the error message `invalid type: found string "{certs=’/ssl/cert1.pem',key='/ssl/fullchain1.pem'}", expected struct TlsConfig` and `invalid type: found string "'{certs="/ssl/fullchain.pem",key="/ssl/privkey.pem"}'", expected struct TlsConfig`. I.e. Rocket does not even try to load the files.

@bmpreston I don't know how you configure your environment variables that some of them end up having single quotes around them
![grafik](https://github.com/dani-garcia/vaultwarden/assets/509385/2cd152ee-9900-491e-8097-a188b3952770)
 but like I already said, try to just set the value to `{certs="/ssl/fullchain.pem",key="/ssl/privkey.pem"}`.

**Reply by bmpreston** [reply]
<img width="780" alt="Screenshot" src="https://github.com/dani-garcia/vaultwarden/assets/29456939/1ad50dd9-0609-4057-9c3a-c04d92f0174e">
<img width="925" alt="Screenshot" src="https://github.com/dani-garcia/vaultwarden/assets/29456939/e91b9d20-1de5-4af3-afbf-9f66f5ad0d1c">

DIFFERENT!!!

The actual mid-value quote worked.  I missed the QUOTE in the middle of the declaration.  

<img width="954" alt="Screenshot" src="https://github.com/dani-garcia/vaultwarden/assets/29456939/b012c96d-c26b-47c8-aeee-234a6b2ecd28">

**Reply by stefan0xC** [reply]
Now you just have to fix the files you use (i.e. use `fullchain1.pem` as certs and `privkey1.pem` as key).

**Reply by bmpreston** [reply]
Yeah, I'm researching to figure out what those errors mean.

**Reply by stefan0xC** [reply]
It means you provided the wrong file (e.g. a certificate, where Rocket expected a key).

**Reply by bmpreston** [reply]
Thank you again!
